OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs

```diff
@@ -117,6 +120,28 @@ IEnumerable<Order> InnerOrder(World world, CPos cell, MouseInput mi)
 						yield break;
 					}
 				}
+				else if (replacmentInfo != null)
+				{
+					var res = world.WorldActor.TraitOrDefault<ResourceLayer>();
+					var occupied = false;
+					foreach (var t in buildingInfo.Tiles(topLeft))
+						if (!occupied) occupied = !(world.IsCellBuildable(t, buildingInfo)
+							|| AcceptsReplacement(t, replacmentInfo))
+							|| !(res == null || res.GetResource(t) == null);
+
+					if (!buildingInfo.IsCloseEnoughToBase(world, owner, building, topLeft)
+						|| occupied)
+					{
+						foreach (var order in ClearBlockersOrders(world, topLeft))
+							yield return order;
+
+						Game.Sound.PlayNotification(world.Map.Rules, owner, "Speech", "BuildingCannotPlaceAudio", owner.Faction.InternalName);
+						yield break;
+					}
+
+					if (world.Map.Rules.Actors[building].HasTraitInfo<ReplacementInfo>() && !mi.Modifiers.HasModifier(Modifiers.Shift))
```

Comment:
The `HasTraitInfo` check shouldn't be needed anymore if I'm not mistaken, as you already check if `replacementInfo != null`.